---
title: "Detection of k-mers associated to bioclim variables"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


This work is mainly done on IFB cluster.


## Objectives
The purpose is to detect kmers associated with bioclim variables (and maybe also elevation data), using the native samples in Africa.    
Firstly, the Kiss pipeline will be applied to obtain a kmer data set, and detect kmers under selection (associated with population structure, that will be useful for the association test).   
Secondly, I will apply lfmm with different strategies (the whole kmer data set, the putatively adaptive kmers, putatively adaptive kmers + neutral kmers). Then, I will check the distribution of the p-values and/or q-values to determine which strategy is better in terms of minimizing false positives. 
Finally, the significant kmers will be mapped on the reference genome and assembled for annotations ... (to be developed)

## Materials

### Previously sequenced samples

55 African individuals were sequenced using Illumina.  
They are representative of the 8 genetic groups and consist of some admixed individuals.  

They were cleaned by filtering adapters and trimming low-quality read bases.

removing adapters by fastp 0.23.1: 
`fastp --detect_adapter_for_pe --disable_quality_filtering --disable_length_filtering`  
trimming read bases < Q20 (min read length = 35) by cutadapt 3.1: 
`cutadapt -m 35 -q 20,20`

The clean reads are available in `/data/projects/rob/data/raw_fastq_2019` on iTrop,
and `/shared/projects/elai_most/data/fastq/fastq_2019_clean` on IFB.

### Newly sequenced samples

We selected 20 African samples with more pure genetic structure based on snmf results using kasp SNPs.    
These samples were degraded (level B or C on BGI quality report, and so some digestion step was skipped). 
Therefore, we need to check carefully the sequecing quality.

```{bash rename_fastq}
fastqfiles=$(ls /shared/projects/afrob_seq/fastq_2022/Reads/Clean/*/*)
newpath=/shared/projects/afrob_seq/fastq_2022/renamed
mkdir $newpath
for f in $fastqfiles
do
    sample=$(echo $(basename $(dirname $f)) | sed 's/A//')
    read=$(echo ${f:(-7)})
    mkdir -p $newpath/$sample
    ln -s $f $newpath/$sample/$sample"_"$read
done
ll $newpath
```

```{bash quality_control}
mkdir -p /shared/projects/afrob_seq/fastq_2022/quality
cd /shared/projects/afrob_seq/fastq_2022/quality
nano qc_config.yaml
nano run_qc.sh
sbatch run_qc.sh
```

#!/bin/bash
#SBATCH --job-name qc_2022
#SBATCH --partition=long
#SBATCH --cpus-per-task=1
#SBATCH --output /shared/projects/afrob_seq/fastq_2022/quality/slurm-%x_%j.log
#SBATCH --error /shared/projects/afrob_seq/fastq_2022/quality/slurm-%x_%j.log

module load singularity
module load conda
module load snakemake/6.12.3

cd /shared/projects/afrob_seq/wgs_snp/scripts/gbp_variantcalling

snakemake fastqc --use-conda -p --verbose --rerun-incomplete \
--jobs 22 --latency-wait 30 --nolock --notemp \
--configfile /shared/projects/afrob_seq/fastq_2022/quality/qc_config.yaml \
--cluster "python3 slurm_wrapper.py" \
--cluster-status "python3 slurm_status.py" 



```{bash}
cd /shared/projects/afrob_seq/wgs_snp/scripts/gbp_variantcalling
```

