---
title: "kmer analysis"
author: "Tram"
date: "6/13/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(BEDMatrix)
library(readr)
library(LEA)
library(vcfR)
library(tidyverse)
library(data.table)
library(lfmm)
library(psych)
library(car)
library(GGally)
library(viridis)
library(vegan)
library(ggfortify)
```

```{r, eval=FALSE}
setwd("/data/projects/rob/scripts/kmer")
```

`k-mers can identify not only almost all associations found by SNPs and short indels but also SVs and variants in sequences not present in reference genomes`

# Objectives:

1.  detect k-mers under selection (associated with population structure)
2.  detect k-mers associated to local climate

# Methods:

1.  run Kiss on the african set (55 inds) -> confirm genetic structure with SNPs using snmf

2.  run Kiss on the whole set (55 AF + 45 VN) -> select k-mers using 2 statistical methods

3.  GEA

## Test Kiss on the african set


### On I-trop

I had to run it on /scratch, and the results were automatically deleted after few weeks...

```{bash}
sbatch /data/projects/rob/scripts/kiss_scripts/af_kiss_ci2.sh
```

first I tried pcadapt with K=2\
then I found from the scree plot that K = 7 should be a better value\
because it is the largest value before plateau and it is congruent with the cross entropy evaluation by snmf


### On IFB

```{bash}
# transfer fastq data 
cd /shared/projects/most_kmer/
mkdir -p kiss_af/fastq
cd kiss_af
rsync -vauP baotram@bioinfo-nas2.ird.fr:/data/projects/rob/data/raw_fastq_2019/* fastq/
```


### choose a random k-mer set

```{r}
set.seed(123)
rand_set <- sample(1:1728, 1) # 415
bed_nb <- rand_set %/% 100 # 4
range_nb <- rand_set %% 100 # 15

scratch_dir <- "/scratch/kmer/af_kiss/results_ci2"
kiss_dir <- "./tmp/af_ci2"
kiss_dir <- "/shared/projects/most_kmer/kiss_out"
```

-> batch 4, subset 15

### download the corresponding bed file and kmer list file from /scratch

```{bash}
ssh node25
rsync -vaurRP /scratch/kmer/af_kiss/results_ci2/./3.TABLE2BED/output_file.4.* /data/projects/rob/scripts/kmer/tmp/af_ci2
rsync -vaurRP  /scratch/kmer/af_kiss/results_ci2/./5.RANGES/output_file.4/15.txt /data/projects/rob/scripts/kmer/tmp/af_ci2
rsync -vaurRP  /scratch/kmer/af_kiss/results_ci2/./6.PCADAPT/output_file.4_15_BH0.05.pcadapt.rplot.pdf /data/projects/rob/scripts/kmer/tmp/af_ci2
```

### read the files

```{r}
bed_file <- file.path(kiss_dir, "3.TABLE2BED/output_file.4.bed")
matrix <- BEDMatrix(bed_file)

# bim_file <- file.path(kiss_dir, "3.TABLE2BED/output_file.4.bim")
# bim <- read_delim(bim_file, delim = "\t", col_names = F)

kmer_file <- file.path(kiss_dir, "5.RANGES/output_file.4/15.txt")
kl <- scan(kmer_file)

kmer_names_list <- sort(kl) # kmer number by segment
submat <- matrix[, sort(kl)]

af_sample_file <- "./af_sample_name.txt"
af_sample <- read.delim(af_sample_file)

af_kiss_file <- "./af_sample.txt"
af_kiss <- read.delim(af_kiss_file)

af_sample <- left_join(af_kiss %>% select(accession_id), af_sample)
```

### run snmf on the randomly selected kmer set

```{r}
## SNMF
snmf_dir <- file.path(tmp_dir, "snmf")
dir.create(snmf_dir, showWarnings = F)
geno_file <- file.path(snmf_dir, "kmer.geno")
write.geno(submat, geno_file)

snmf_kmer <- snmf(geno_file, K = 1:10, CPU = 2, repetitions = 10, project = "new", entropy = T, seed = 987654321)

plot(snmf_kmer, pch = 16)

## best k = 5
## choose best run based on cross entropy
k <- 5
best_run <- which.min(cross.entropy(snmf_kmer, K = k))

## plot ancestry proportion
ind_order <- barchart(snmf_kmer, K = k, run = best_run, plot = F)
kmer_ref_prop <- as.data.frame(Q(snmf_kmer, K = k, run = best_run))
kmer_ref_prop$Label <- paste(af_sample$accession_id, af_sample$phenotype_value, sep = "_")
kmer_ref_prop$prior <- factor(af_sample$phenotype_value, levels = c("ER", "OB", "C", "AG", "D"))
kmer_ref_prop <- pivot_longer(kmer_ref_prop, cols = V1:V5, names_to = "ancestry", values_to = "proportion")
## rename ancestry
repre <- kmer_ref_prop %>% group_by(ancestry) %>% arrange(desc(proportion)) %>% slice_head(n = 1) %>% slice_tail(n = 1)
# rename_ancestr <- sapply(repre$ancestry, function(i) af_sample$phenotype_value[af_sample$accession_id == repre$Label[repre$ancestry == i]])
# levels(prop$ancestry) <- repre$prior

kmer_ref_prop$Label <- factor(kmer_ref_prop$Label, levels = unique(kmer_ref_prop$Label)[ind_order$order], ordered = F)
kmer_ref_prop$ancestry <- sapply(kmer_ref_prop$ancestry, function(i) repre$prior[repre$ancestry == i], USE.NAMES = F)

ggplot(kmer_ref_prop) +
  geom_col(aes(x = Label, y = proportion, fill = ancestry), position = position_stack(), width = .95, size = 0) + 
  # scale_fill_viridis_d(end = .9)+
  scale_fill_manual(values = turbo(5)) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_continuous(expand = c(0, 0)) +
  ylab("Ancestry proportion")  +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, size = 10, hjust = 1),
    axis.title.y = element_text(size = 12),
    # axis.title.x = element_blank(),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 12),
    legend.key.size = unit(5, "mm"),
    axis.ticks.y = element_line(size = .1),
    axis.text.y = element_text(size = 8),
    panel.spacing = unit(2, "mm"),
    # strip.text.x = element_text(angle = 90, size = 10, hjust = 1)
    ) 
```

```{r}
k <- 6
best_run <- which.min(cross.entropy(snmf_kmer, K = k))

## plot ancestry proportion
ind_order <- barchart(snmf_kmer, K = k, run = best_run, plot = F)
kmer_ref_prop <- as.data.frame(Q(snmf_kmer, K = k, run = best_run))
kmer_ref_prop$Label <- paste(af_sample$accession_id, af_sample$label, sep = "_")
kmer_ref_prop$Label <- factor(kmer_ref_prop$Label, levels = unique(kmer_ref_prop$Label)[ind_order$order], ordered = F)
kmer_ref_prop <- pivot_longer(kmer_ref_prop, cols = V1:V6, names_to = "ancestry", values_to = "proportion")

ggplot(kmer_ref_prop) +
  geom_col(aes(x = Label, y = proportion, fill = ancestry), position = position_stack(), width = .95, size = 0) + 
  # scale_fill_viridis_d(end = .9)+
  scale_fill_manual(values = turbo(7)) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_continuous(expand = c(0, 0)) +
  ylab("Ancestry proportion")  +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, size = 10, hjust = 1),
    axis.title.y = element_text(size = 12),
    # axis.title.x = element_blank(),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 12),
    legend.key.size = unit(5, "mm"),
    axis.ticks.y = element_line(size = .1),
    axis.text.y = element_text(size = 8),
    panel.spacing = unit(2, "mm"),
    # strip.text.x = element_text(angle = 90, size = 10, hjust = 1)
    ) 
```

### run snmf on window-based SNPs

```{bash}
rsync -vaurP baotram@core.cluster.france-bioinformatique.fr:/shared/projects/most_kmer/vcf_bgi_2021/gbp_output/filtered_vcf_by_chrom/all_final_thin5000_maf002.recode.vcf /data/projects/rob/scripts/kmer/tmp

rsync -vaurP /scratch/kmer/all_kiss/6.PCADAPT/output_file.10_3_BH0.05.pcadapt.rplot.pdf  /data/projects/rob/scripts/kmer/tmp/all
```

```{r}
vcf_file <- "/data/projects/rob/scripts/kmer/tmp/all_final_thin5000_maf002.recode.vcf"
vcf_header <- scanBcfHeader(vcf_file)
ind_names <- vcf_header[[1]]@listData$Sample
ref_names <- grep("(S-|TR)", ind_names, invert = T, value = T)
ref_names <- sapply(ref_names, function(i) {
  r <- ifelse(grepl("NoIndex", i),
         gsub("(^C|_.+)", "", i),
         gsub("_.+", "", i)
         )
  # r <- gsub("_.+", "", i)
  # return(r)
  id <- grep(r, af_sample$accession_id, value = T)
  prior <- af_sample$phenotype_value[af_sample$accession_id == id]
  return(paste(id, prior, sep = "_"))
}, USE.NAMES = F)
```

```{r}
snp_snmf <- load.snmfProject("/data/projects/rob/data/most_elai/snmf/thin_ref.snmfProject")
ind_order_snp <- barchart(snp_snmf, K = k, run = which.min(cross.entropy(snp_snmf, K = 5)), plot = F)
snp_ref_prop <- as.data.frame(Q(snp_snmf, K = k, run = which.min(cross.entropy(snp_snmf, K = k))))
snp_ref_prop$Label <- ref_names
snp_ref_prop$prior <- factor(gsub(".+_", "", ref_names), levels = c("ER", "OB", "C", "AG", "D"))
snp_ref_prop <- pivot_longer(snp_ref_prop, cols = V1:V5, names_to = "ancestry", values_to = "proportion")
## rename ancestry
repre_snp <- snp_ref_prop %>% group_by(ancestry) %>% arrange(desc(proportion)) %>% slice_head(n = 1) %>% slice_tail(n = 1)
# rename_ancestr <- sapply(repre$ancestry, function(i) af_sample$phenotype_value[af_sample$accession_id == repre$Label[repre$ancestry == i]])
# levels(prop$ancestry) <- repre$prior

snp_ref_prop$Label <- factor(snp_ref_prop$Label, levels = unique(snp_ref_prop$Label)[ind_order_snp$order], ordered = F)
snp_ref_prop$ancestry <- sapply(snp_ref_prop$ancestry, function(i) repre_snp$prior[repre_snp$ancestry == i], USE.NAMES = F)

ggplot(snp_ref_prop) +
  geom_col(aes(x = Label, y = proportion, fill = ancestry), position = position_stack(), width = .95, size = 0) + 
  # scale_fill_viridis_d(end = .9)+
  scale_fill_manual(values = turbo(5)) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_continuous(expand = c(0, 0)) +
  ylab("Ancestry proportion")  +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, size = 10, hjust = 1),
    axis.title.y = element_text(size = 12),
    # axis.title.x = element_blank(),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 12),
    legend.key.size = unit(5, "mm"),
    axis.ticks.y = element_line(size = .1),
    axis.text.y = element_text(size = 8),
    panel.spacing = unit(2, "mm"),
    # strip.text.x = element_text(angle = 90, size = 10, hjust = 1)
    ) 
```

### compare snmf results of 1M kmers and the snps set

```{r}
both_prop <- full_join(kmer_ref_prop %>% mutate(type = "1M KMER"),
          snp_ref_prop %>% mutate(type = "100K SNP"))
pdf("/data/projects/rob/data/kmers/plots/structure_check.pdf", width = 15, height = 8)
ggplot(both_prop) +
  facet_grid(rows = vars(type)) +
  geom_col(aes(x = Label, y = proportion, fill = ancestry), position = position_stack(), width = .95, size = 0) + 
  # scale_fill_viridis_d(end = .9)+
  scale_fill_manual(values = turbo(5)) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_continuous(expand = c(0, 0)) +
  ylab("Ancestry proportion")  +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, size = 10, hjust = 1),
    axis.title.y = element_text(size = 12),
    axis.title.x = element_blank(),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 12),
    legend.key.size = unit(5, "mm"),
    axis.ticks.y = element_line(size = .1),
    axis.text.y = element_text(size = 8),
    panel.spacing = unit(4, "mm"),
    strip.text.y = element_text(angle = -90, size = 15, hjust = 0.5)
    ) 
dev.off()
```

```{r}
kmer_snp_cor <- full_join(kmer_ref_prop %>% rename(kmer = proportion),
          snp_ref_prop %>% rename(snp = proportion)) %>% 
  group_by(ancestry) %>% 
  summarise(correlation = cor(kmer, snp)) %>% 
  pull(correlation) %>% mean()
cat("correlation between snmf of kmer set and snp set:", kmer_snp_cor)
```

-> very high correlation

### run pca on the snp set

```{r}
af_snp_geno_file <- "/data/projects/rob/data/most_elai/snmf/thin_ref.geno"
af_snp_pca <- pca(af_snp_geno_file)
af_snp_proj <- as.data.frame(af_snp_pca$projections[,1:2])
af_snp_proj$accession_id <- gsub("_.{1,2}$", "", ref_names)
af_snp_proj <- left_join(af_snp_proj, af_sample)
colnames(af_snp_proj)[1:2] <- c("PC1", "PC2")
af_snp_eigen <- af_snp_pca$eigenvalues[,1]
af_snp_proj$label <- factor(af_snp_proj$label, levels = c("A", "B", "C", "D", "E", "G", "O", "R", "hb"))

pdf("/data/projects/rob/data/kmers/plots/af_pca_100kSNPs.pdf")
ggplot(af_snp_proj) +
  # geom_point(aes(x = PC1, y = PC2, color = group), alpha = .5) +
  geom_text(aes(x = PC1, y = PC2, label = label, color = label), show.legend = F, size = 6) +
  xlab(paste0("PC1 (", format(af_snp_eigen[1]/sum(af_snp_eigen)*100, digits = 2), "%)")) +
  ylab(paste0("PC2 (", format(af_snp_eigen[2]/sum(af_snp_eigen)*100, digits = 2), "%)"))  +
  theme_minimal() +
  theme(axis.title = element_text(size = 15), panel.border = element_rect(color = "gray", fill = NA))
dev.off()
```

## Kiss on the whole set

### pca on the snp set

```{r}
ind_names <- sapply(ind_names, function(i) {
    r <- ifelse(grepl("NoIndex", i),
                gsub("(^C|_.+)", "", i),
                gsub("_.+", "", i)
    )
    # r <- gsub("_.+", "", i)
    # return(r)
    id <- grep(r, sample$accession_id, value = T)
    # prior <- sample$phenotype_value[sample$accession_id == id]
    return(id)
}, USE.NAMES = F)
```

```{r}
snp_geno_file <- "/data/projects/rob/data/most_elai/snmf/thin_all.geno"
snp_pca <- pca(snp_geno_file)
snp_proj <- as.data.frame(snp_pca$projections[,1:2])
snp_proj$accession_id <- ind_names
snp_proj <- left_join(snp_proj, sample) %>% 
  rename(PC1 = V1, PC2 = V2) %>% 
  mutate(prior = gsub("(S-|TR)", NA, label))
snp_eigen <- snp_pca$eigenvalues[,1]
snp_proj$prior <- factor(snp_proj$prior, levels = c("A", "B", "C", "D", "E", "G", "O", "R", "hb", NA))


pdf("/data/projects/rob/data/kmers/plots/all_pca_100kSNPs.pdf")
ggplot(snp_proj[nrow(snp_proj):1,]) +
  # geom_point(aes(x = PC1, y = PC2, color = group), alpha = .5) +
  geom_text(aes(x = PC1, y = PC2, label = label, color = prior), show.legend = F, size = 6) +
  xlab(paste0("PC1 (", format(snp_eigen[1]/sum(snp_eigen)*100, digits = 2), "%)")) +
  ylab(paste0("PC2 (", format(snp_eigen[2]/sum(snp_eigen)*100, digits = 2), "%)"))  +
  theme_minimal() +
  theme(axis.title = element_text(size = 15), panel.border = element_rect(color = "gray", fill = NA))
dev.off()
```

-> 399,265,658 adaptive k-mers (K = 7)

## GEA

Below is a trial on a random k-mer set (batch 415) of the african data

### environment variables

```{r}
clim_file <- "./gea/climatic_variables_2.1.csv"
clim <- fread(clim_file, data.table = T)
# write.csv(clim[,1:21], "./gea/lfmm/lfmm_explanatory.csv", 
#           row.names = F, quote = F)
clim <- clim %>% column_to_rownames("Label")

clim <- clim %>% 
  mutate(snp_group = case_when(snp_group %in% c("A", "G") ~ "AG",
                               snp_group %in% c("E", "R") ~ "ER",
                               snp_group %in% c("O", "B") ~ "OB",
                               TRUE ~ snp_group))
```

```{r}
pairs.panels(clim[,1:20], hist.col = "royalblue", ellipses = F)

```

the correlation of the "annual" variables

```{r}
test_clim <- clim[, c(1,7,12)]
pdf("/data/projects/rob/data/kmers/gea/annual_climate_correlation.pdf")
pairs.panels(test_clim, hist.col = "royalblue", ellipses = F)
# legend(1, pch = c(0:5), legend = levels(as.factor(clim$snp_group)))
dev.off()
```

```{r}
svg("/data/projects/rob/data/kmers/gea/climate_correlation_groups.svg")
scatterplotMatrix(test_clim, groups = clim$snp_group)
dev.off()
```

```{r}
test_clim$snp_group <- factor(clim$snp_group, levels = c("ER", "OB", "C", "AG", "D", "hb"))
ggpairs(test_clim, columns = 1:3, aes(color = snp_group)) + 
  scale_color_manual(values = c(turbo(5), "gray")) +
  scale_fill_manual(values = c(turbo(5), "gray")) +
  theme_minimal() +
  theme(panel.background = element_rect(color = "gray"))
ggsave("gea/annual_climate_correlation_groups_gg.svg", width = 8, height = 8)
```

the correlation of the variables maximizing the deviance (Tournebize et al. 2022)

```{r}
test_clim <- clim[, c(2,10,12, 15, 18)]
svg("/data/projects/rob/data/kmers/gea/deviance_climate_correlation.svg")
pairs.panels(test_clim, hist.col = "royalblue", ellipses = F)
# legend(1, pch = c(0:5), legend = levels(as.factor(clim$snp_group)))
dev.off()
```

```{r}
test_clim$snp_group <- factor(clim$snp_group, levels = c("ER", "OB", "C", "AG", "D", "hb"))
ggpairs(test_clim, columns = 1:5, aes(color = snp_group)) + 
  scale_color_manual(values = c(turbo(5), "gray")) +
  scale_fill_manual(values = c(turbo(5), "gray")) +
  theme_minimal() +
  theme(panel.background = element_rect(color = "gray"))
ggsave("gea/deviance_climate_correlation_groups_gg.svg", width = 8, height = 8)
```

distribution of all the variables by group

```{r}
# ggpairs(clim, columns = 1:19, aes(color = snp_group)) + 
#   scale_color_manual(values = c(turbo(5), "gray")) +
#   scale_fill_manual(values = c(turbo(5), "gray")) +
#   theme_minimal() +
#   theme(panel.background = element_rect(color = "gray"))

clim %>% group_by(snp_group) %>% summarise(nb_inds = n())

clim$snp_group <- factor(clim$snp_group, levels = c("hb", "ER", "OB", "C", "AG", "D"))
clim[,c(1:20, ncol(clim))] %>% 
  pivot_longer(cols = 1:20, names_to = "variable") %>% 
  mutate(variable = factor(variable, 
                           levels = str_sort(unique(variable), numeric = T))) %>% 
  filter(snp_group != "hb") %>% 
  rename(group = snp_group) %>% 
  ggplot() +
  facet_wrap(~ variable, ncol = 5, scale = "free") +
  geom_density(aes(value, color = group, fill = group), alpha = 0.5) +
  scale_color_manual(values = turbo(5)) +
  scale_fill_manual(values =  turbo(5)) +
  theme_minimal()
```

```{r}
library(rnaturalearth)
group_col <- c("#FB8022FF", "#A2FC3CFF", "#7A0403FF", "#30123BFF", "#28BBECFF")

africa <- ne_countries(scale = "medium", continent = "africa", returnclass = "sf")


africa <- africa %>% 
  mutate(ancestry = case_when(sovereignt %in% c("Ivory Coast", "Guinea", "Ghana") ~ "D",
                              sovereignt == "Cameroon" ~ "C",
                              sovereignt %in% c("Gabon", "Benin", "Angola") ~ "AG",
                              sovereignt %in% c("Uganda", "Central African Republic") ~ "OB",
                              sovereignt == "Democratic Republic of the Congo" ~ "ER"))

rob_countries <- africa %>% filter(!is.na(ancestry))

ggplot(data = africa) +
  geom_sf(fill = "white", color = "grey") +
  geom_point(data = clim, aes(x = Long, y = Lat, color = snp_group)) +
  stat_density2d(data = clim, aes(x = Long, y = Lat, fill = snp_group), alpha = .5,
                 geom = "polygon") +
  # geom_sf(data = rob_countries, aes(fill = ancestry), color = "grey") +
  # geom_sf_text(data = rob_countries, aes(label = abbrev), color = "grey50") +
  scale_color_manual(values = c(group_col, "gray"), na.value = "grey50") +
  scale_fill_manual(values = c(group_col, "gray"), na.value = "grey50") +
  coord_sf(xlim = c(-15, 48), ylim = c(-40, 35), expand = TRUE) +
  theme_void() +
  theme(legend.box.margin = margin(l = 20), 
        legend.key.size = unit(14, "pt"),
        legend.key.height = unit(15, "pt"))
```

### toy genotype

```{r}
submat1 <- submat
rownames(submat1) <- gsub("_.+", "", rownames(submat1))
submat1 <- submat1[rownames(submat1) %in% rownames(clim),]
```

### lfmm with all variables

```{r}
k <- 5 # based on pcadapt and snmf results
```

```{r}
suball_lfmm <- lfmm_ridge(Y = submat1, X = clim[, 1:20], K = k)
pv <- lfmm_test(Y = submat1, X = clim[, 1:20], 
                lfmm = suball_lfmm, 
                calibrate = "median+MAD")
hist(pv$calibrated.pvalue)

test_lffm_res <- "./gea/test_lfmm_results.pdf"
pdf(test_lffm_res)
for (i in 1:ncol(clim[, 1:20])) {
  pvalues <- pv$calibrated.pvalue[,i]
  # hist(pvalues, 
  #      main = paste0("Calibrated p-values for ", colnames(clim[, 1:20])[i], 
  #                    " (GIF = ", sprintf("%.2f", pv$gif[i]), ")"),
  #      xlab = "p-values")
  qqplot(rexp(length(pvalues), rate = log(10)),
       -log10(pvalues), 
       xlab = "Expected quantile", main = colnames(clim[, 1:20])[i], 
       pch = 19, cex = .4)
  abline(0,1)
}
dev.off()
```

```{r}
associated_kmers <- rownames(pv$calibrated.pvalue)[sapply(1:nrow(pv$calibrated.pvalue), function(i) any(pv$calibrated.pvalue[i, 1:19] < 0.05))]
length(unique(associated_kmers))

u <- 0
for (i in 1:19) {
  cat(paste0("bio", i, ": "))
  v <- length(rownames(pv$calibrated.pvalue)[pv$calibrated.pvalue[,i] < 0.05])
  u <- u + v
  cat(v)
  cat("\n")
  }
```

-> 507468 significant k-mers -> the histogram seems fine but half of the k-mers are significant

### lfmm with only annual climatic variables

```{r}
subann_lfmm <- lfmm_ridge(Y = submat1, X = clim[, c(1,3,7,12)], K = k)
pv_ann <- lfmm_test(Y = submat1, X = clim[, c(1,3,7,12)], 
                    lfmm = subann_lfmm, 
                    calibrate = "gif")
hist(pv_ann$calibrated.pvalue)

signif_ann <- rownames(pv_ann$calibrated.pvalue)[sapply(1:nrow(pv_ann$calibrated.pvalue), function(i) any(pv_ann$calibrated.pvalue[i,] < 0.05))]
length(unique(signif_ann))
```

-> 263789 significant k-mers

### lfmm with pca axes

```{r}
clim_pca <- rda(clim[,1:19], scale = T)
summary(clim_pca)$cont
round(scores(clim_pca, choices = 1:2, display = "species", scaling = 0), digits = 3)
biplot(clim_pca)
```

```{r}
clim_pure <- clim[clim$snp_group != "hb",]
clim_pca2 <- prcomp(clim_pure[,1:19], scale = T)
summary(clim_pca2)
screeplot(clim_pca2)
library(ggfortify)
clim_pure$snp_group <- factor(clim_pure$snp_group, levels = c("ER", "OB", "C", "AG", "D"))
## PC1 vs PC2
autoplot(clim_pca2, x = 1, y = 2,
         data = clim_pure, colour = 'snp_group', size = 2,
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 6,
         loadings.label.colour = "black", frame = T) +
  theme_minimal() + 
  scale_colour_manual(values = c(turbo(5))) +
  scale_fill_manual(values = c(turbo(5))) 

## PC1 vs PC2
autoplot(clim_pca2, x = 1, y = 3,
         data = clim_pure, colour = 'snp_group', size = 2,
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 6,
         loadings.label.colour = "black", frame = T) +
  theme_minimal() + 
  scale_colour_manual(values = c(turbo(5))) +
  scale_fill_manual(values = c(turbo(5))) 
  
```

```{r}
test_clim_pca <- clim[, c(12,14,17)]
svg("/data/projects/rob/data/kmers/gea/climate_correlation.svg")
pairs.panels(test_clim_pca, hist.col = "royalblue", ellipses = F, pch = c(0:5)[as.factor(clim$snp_group)])
legend(1, pch = c(0:5), legend = levels(as.factor(clim$snp_group)))
dev.off()

ggpairs(clim, columns = c(12,14,17), aes(color = snp_group)) + 
  scale_color_manual(values = c(turbo(5), "gray")) +
  scale_fill_manual(values = c(turbo(5), "gray")) +
  theme_minimal() +
  theme(panel.background = element_rect(color = "gray"))

ggpairs(clim, columns = c(1,5,8:10), aes(color = snp_group)) + 
  scale_color_manual(values = c(turbo(5), "gray")) +
  scale_fill_manual(values = c(turbo(5), "gray")) +
  theme_minimal() +
  theme(panel.background = element_rect(color = "gray"))
```

```{r}
clim_PC1 <- scores(clim_pca, choices = 1, display = "sites", scaling = 0)
subpca_lfmm <- lfmm_ridge(Y = submat1, X = clim_PC1, K = k)

pv_pca <- lfmm_test(Y = submat1, X = clim_PC1, 
                    lfmm = subpca_lfmm, 
                    calibrate = "gif")
hist(pv_pca$calibrated.pvalue)
pv_pca$gif

signif_pca <- rownames(pv_pca$calibrated.pvalue)[sapply(1:nrow(pv_pca$calibrated.pvalue), function(i) any(pv_pca$calibrated.pvalue[i,] < 0.05))]
length(unique(signif_pca))
```

```{r}
clim_PC2 <- scores(clim_pca, choices = 1:2, display = "sites", scaling = 0)
subpca2_lfmm <- lfmm_ridge(Y = submat1, X = clim_PC2, K = k)

pv_pca2 <- lfmm_test(Y = submat1, X = clim_PC2, 
                    lfmm = subpca2_lfmm, 
                    calibrate = "gif")
hist(pv_pca2$calibrated.pvalue)
for (i in 1:2) {
  hist(pv_pca2$calibrated.pvalue[,i], 
       main = paste0("Calibrated p-values for PC", i, 
                     " (GIF = ", sprintf("%.2f", pv_pca2$gif[i]), ")"),
       xlab = "p-values")
}
pv_pca2$gif

signif_pca2 <- rownames(pv_pca2$calibrated.pvalue)[sapply(1:nrow(pv_pca2$calibrated.pvalue), function(i) any(pv_pca2$calibrated.pvalue[i,] < 0.05))]
length(unique(signif_pca2))

plot(-log10(pv_pca2$calibrated.pvalue[,1]), 
      pch = 19, 
      cex = .2, 
      xlab = "k-mer", ylab = "-Log P",
      col = "grey")
plot(-log10(pv_pca2$calibrated.pvalue[,2]), 
      pch = 19, 
      cex = .2, 
      xlab = "k-mer", ylab = "-Log P",
      col = "grey")
```

-> 144573 k-mers

### genome-wide lfmm using pca of clim vars

gather all the lfmm results for all the k-mer batches

```{r}
lfmm_out <- "/data/projects/rob/scripts/kmer/gea/lfmm/out"
lfmm_files <- list.files(lfmm_out, pattern = ".*.csv", full.names = T, recursive = T)
all_lfmm <- file.path(lfmm_out, "all_candidates.csv")
file.append(all_lfmm, lfmm_files)
```

count the number of total candidate kmers

```{bash}
wc -l /data/projects/rob/scripts/kmer/gea/lfmm/out/all_candidates.csv
```

-> 218,269,911 k-mers (12%)

### genome-wide lfmm using all the 19 clim vars

we decided to use all the 19 variables since we can associate k-mers to each climatic variable

for working on IFB: transfer kiss results from Itrop to IFB\
(because running 1728 independent jobs on IFB is much x3000 faster)

```{bash update_ifb}
cd /shared/projects/most_kmer
mkdir kiss_out
rsync -vaurP baotram@bioinfo-nas2.ird.fr:/data/projects/rob/scripts/kmer/gea/tmp/* kiss_out/
mkdir -p /shared/projects/most_kmer/afterkiss/gea/lfmm/log
sbatch /shared/projects/most_kmer/afterkiss/gea/lfmm/run_lfmm_af_IFB.sh
```

```{bash}
ssh baotram@core.cluster.france-bioinformatique.fr
module load singularity
singularity exec -B /shared/home/baotram \
/shared/projects/vietcaf/rserver/Singularity.R_4-2-0-Rserver_2022.sif \
Rscript /shared/projects/most_kmer/afterkiss/gea/lfmm/install_packages.R
```

then, transfer lfmm results from IFB to Itrop

```{bash}
rsync -vaurP baotram@core.cluster.france-bioinformatique.fr:/shared/projects/most_kmer/afterkiss/gea/lfmm/out_K5 /data/projects/rob/scripts/kmer/gea/lfmm/
rsync -vaurP baotram@core.cluster.france-bioinformatique.fr:/shared/projects/most_kmer/afterkiss/gea/lfmm/log/* /data/projects/rob/scripts/kmer/gea/lfmm/log_K5
```

check if all the runs have finished

```{r}
lfmm_out <- "./gea/lfmm/out_K5"
real <- dirname(list.files(lfmm_out, "lfmm_results.pdf", full.names = F, recursive = T))
length(real)

n <- 1728
expect <- sapply(1:n, function(r) {
  if (r %% 100 == 0) {
    b <- r/100 - 1
    s <- 100
  } else {
    b <- r %/% 100
    s <- r %% 100
  }
  return(paste(b, s, sep = "_"))
})
length(expect)

# files that haven't finished
expect[!expect %in% real]
```

gather all the lfmm results for all the k-mer batches using the results directly

```{r}
all_lfmm_dir <- "./gea/lfmm/all_bio"
dir.create(all_lfmm_dir, showWarnings = T)
lfmm_files <- list.files(lfmm_out, pattern = "_bio1.csv", full.names = T, recursive = T)
all_lfmm <- file.path(all_lfmm_dir, "all_candidates_bio1.csv")
file.append(all_lfmm, lfmm_files)

for (i in 1:19) {
  lfmm_files <- list.files(lfmm_out, pattern = paste0("_bio", i, ".csv"), 
                           full.names = T, recursive = T)

}

## below: it works but very slow and memory consuming
library(R.utils)

cand_kmers <- rowSums(do.call(cbind, lapply(expect, function(i) {
  cand_files <- list.files(file.path(lfmm_out, i), "candidates_.+.csv", full.names = T)
  do.call(rbind, lapply(cand_files, function(f) {
    varname <- gsub("(.+_|.csv)", "", basename(f))
    df <- data.frame(countLines(f))
    colnames(df) <- which(f %in% cand_files)
    rownames(df) <- varname
    return(df)
    }))
})))
```

gather all the lfmm results for all the k-mer batches using the log files (faster i think)

```{r}
log_K5 <- "./gea/lfmm/log_K5"
log_files <- list.files(log_K5, pattern = ".log", full.names = T)
n <- length(log_files)

cand_kmers <- rowSums(do.call(cbind, lapply(1:n, function(i) {
  cat(i, "\n")
  df <- fread(log_files[i], skip = 10, nrows = 20, sep = " ", 
              select = c(7, 9), col.names = c("var", i),
              data.table = FALSE)
  df <- df %>% column_to_rownames("var")
  if (rownames(df)[20] == "elevation") {cat(i, "\n")}
  return(df)
  })
  ))
cand_kmers <- format(as.data.frame(cand_kmers), big.mark = ",")
```

the histograms and qq-plots are not good (stares, uniform distribution)




### lfmm of putative adaptive kmers

so, we try to run lfmm only for the putative adaptive kmers to see if we have better p-values distribution  
for each batch, run pcadapt to select adaptive kmers, then run lfmm for only these adaptive ones.


```{r}
lfmm_out <- "./gea/lfmm/out_K5_outlier/"
real <- dirname(list.files(lfmm_out, "lfmm_results.pdf", full.names = F, recursive = T))
length(real)

n <- 1728
expect <- sapply(1:n, function(r) {
  if (r %% 100 == 0) {
    b <- r/100 - 1
    s <- 100
  } else {
    b <- r %/% 100
    s <- r %% 100
  }
  return(paste(b, s, sep = "_"))
})
length(expect)

# files that haven't finished
expect[!expect %in% real]
```

```{r}
library(R.utils)

cand_kmers <- rowSums(do.call(cbind, lapply(expect, function(i) {
  cand_files <- list.files(file.path(lfmm_out, i), "candidates_.+.csv", full.names = T)
  do.call(rbind, lapply(cand_files, function(f) {
    varname <- gsub("(.+_|.csv)", "", basename(f))
    df <- data.frame(countLines(f))
    colnames(df) <- which(f %in% cand_files)
    rownames(df) <- varname
    return(df)
    }))
})))
```

### RDA

```{r}
sub_rda <- rda(submat1 ~ ., data = clim[, 1:20], scale = T)
summary(sub_rda)$concont
RsquareAdj(sub_rda)
screeplot(sub_rda)
vif.cca(sub_rda)
plot(sub_rda, scaling = 3)
```

```{r}
load_rda <- summary(sub_rda)$species[,1:2]
hist(load_rda[,1])
hist(load_rda[,2])
```

```{r}
outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

cand1 <- outliers(load_rda[,1], 2)
length(cand1)
cand2 <- outliers(load_rda[,2], 2)
length(cand2)
signif_rda <- unique(names(c(cand1, cand2)))
```

-> 132070 candidate k-mers

```{r}
length(signif_pca2[signif_pca2 %in% signif_rda])
length(signif_rda[signif_rda %in% signif_pca2])
```

-> only 40835 common k-mers

## Climatic condition in Central Highlands VN

- Following [this paper](https://doi.org/10.3389/fenvs.2022.820916): select a random point in each district, and get bioclim variables.
    + How to get the coordinates? get the map of the Central Highlands (by raster package) then extract the mid point of each district  
    -> there are 62 districts in the central highlands

- Then, run PCA of bioclim variables (with and without elevation) to see how they are different from the condition in Africa.

### Coordinates of all districts in the Central Highlands
```{r}
vn <- getData("GADM", country = "VN", level = 2)

ch_provinces <- c("Lâm Đồng", "Đắk Lắk", "Đắk Nông", "Kon Tum", "Gia Lai")
ch_map <- vn[vn$NAME_1 %in% ch_provinces, ]
length(ch_map$NAME_2)

chdist_coord <- do.call(rbind, lapply(ch_map$NAME_2, function(d) {
  dist_map <- ch_map[ch_map$NAME_2 == d,]
  coord <- data.frame(province = dist_map$NAME_1,
                      district = dist_map$NAME_2,
                      coordinates(dist_map)
                      )
  colnames(coord)[3:4] <- c("long", "lat")
  return(coord)
}))
```


### Bioclim variables of the districts

```{r}
chvar <- do.call(rbind, lapply(1:nrow(chdist_coord), function(i) {
  clim <- getData(name = "worldclim", var = "bio", res = 0.5, lat = na.omit(chdist_coord$lat)[i], lon = na.omit(chdist_coord$long)[i])
  elev <- getData(name = "SRTM", lat = na.omit(chdist_coord$lat)[i], lon = na.omit(chdist_coord$long)[i])
  point <- SpatialPoints(data.frame(long = na.omit(chdist_coord$long)[i], lat = na.omit(chdist_coord$lat)[i]), proj4string = clim@crs)
  clim_values <- extract(clim, point)
  elev_value <- extract(elev, point)
  colnames(clim_values) <- gsub("_.+", "", colnames(clim_values))
  df <- data.frame(district = chdist_coord$district[i], 
                   clim_values, 
                   elevation = elev_value,
                   Long = chdist_coord$long[i], Lat = chdist_coord$lat[i],
                   group = "VN")
  return(df)
}))
chvar <- chvar %>% 
  column_to_rownames("district")
```


```{r}
all_clim <- getData(name = "worldclim", var = "bio", res = 2.5)

vn_clim <- raster::mask(all_clim, ch_map)
plot(vn_clim[[1:19]], ylim = c(11, 16), xlim = c(105, 110), axes = F, 
     col = viridis::cividis(100), add = T)
```


### Distribution of clim variables in Africa + VN

```{r}
clim_all <- rbind(chvar,
      clim %>% rename(group = snp_group))

clim_groups <- clim_all %>% 
  filter(group != "hb")

clim_groups$group <- factor(clim_groups$group, levels = c("ER", "OB", "C", "AG", "D", "VN"))

clim_groups %>% 
  filter(group != "hb") %>%  
  pivot_longer(cols = 1:20, names_to = "variable") %>% 
  mutate(variable = factor(variable, 
                           levels = str_sort(unique(variable), numeric = T))) %>% 
  ggplot() +
  facet_wrap(~ variable, ncol = 5, scale = "free") +
  geom_density(aes(value, color = group, fill = group), alpha = 0.5) +
  scale_color_manual(values = c(turbo(5), "black")) +
  scale_fill_manual(values = c(turbo(5), "black")) +
  theme_minimal()
```

### PCA of clim variables in Africa+VN
```{r}
climall_pca <- prcomp(clim_groups[,1:20], scale = T)
summary(climall_pca)
screeplot(climall_pca)

## PC1 vs PC2
autoplot(climall_pca, x = 1, y = 2,
         data = clim_groups, colour = 'group', size = 2,
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 6,
         loadings.label.colour = "black", frame = T) +
  theme_minimal() + 
  scale_colour_manual(values = c(turbo(5), "black")) +
  scale_fill_manual(values = c(turbo(5), "black")) 

## PC1 vs PC3
autoplot(climall_pca, x = 1, y = 3,
         data = clim_groups, colour = 'group', size = 2,
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 6,
         loadings.label.colour = "black", frame = T) +
  theme_minimal() + 
  scale_colour_manual(values = c(turbo(5), "black")) +
  scale_fill_manual(values = c(turbo(5), "black")) 
```


```{r}
climall_pca <- prcomp(clim_groups[,1:19], scale = T)
summary(climall_pca)
screeplot(climall_pca)

## PC1 vs PC2
autoplot(climall_pca, x = 1, y = 2,
         data = clim_groups, colour = 'group', size = 2,
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 6,
         loadings.label.colour = "black", frame = T) +
  theme_minimal() + 
  scale_colour_manual(values = c(turbo(5), "black")) +
  scale_fill_manual(values = c(turbo(5), "black")) 

## PC1 vs PC2
autoplot(climall_pca, x = 1, y = 3,
         data = clim_groups, colour = 'group', size = 2,
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 6,
         loadings.label.colour = "black", frame = T) +
  theme_minimal() + 
  scale_colour_manual(values = c(turbo(5), "black")) +
  scale_fill_manual(values = c(turbo(5), "black")) 
```